### PR TITLE
[Run2_2017] Automatically create Docker images as part of a CI GitHub Action

### PR DESCRIPTION
**Pull request by aperloff**
Setup a system whereby a couple of Docker images are setup using a GitHub Action. Both images will have the TreeMaker software installed. One of them will have the CVMFS cache cleared while the other will not. More information about the images can be found in the [TreeMaker wiki](https://github.com/TreeMaker/TreeMaker/wiki/Docker%20Integration%20Test).

### CODE REVIEW

File: .github/workflows/dockerimage.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    - name: Commit the changes
+      run: docker commit -c 'ENTRYPOINT ["/run.sh"]' -c 'CMD []' treemaker treemaker/treemaker:${{ env.current_branch }}-latest
+    - name: Log into registry
+      run: echo "${{ secrets.DOCKER_PASSWORD }}" | docker login -u ${{ secrets.DOCKER_USERNAME }} --password-stdin
+    - name: Publish the new docker image
+      run: docker push treemaker/treemaker:${{ env.current_branch }}-latest
+  integration-test:
+    runs-on: ubuntu-latest
+    needs: [build]
+    if: "!contains(github.event.head_commit.message, '[skip test]')"
```
Comment (kpedro88) [inline .github/workflows/dockerimage.yml:32]: can you update the readme so this is documented somewhere accessible?

File: Production/test/unitTest.py [hunk 2]
```diff
@@ -1,92 +1,157 @@
...
+        print "\nCleaning output files..."
+        files=[self.getROOTName(),self.getLogName()]
+        if clean>0: files=files[0:clean]
+        for file in files:
+            if os.path.exists(file):
+                os.remove(file)
+        print "\nDone cleaning files!"
+
+def defineTests(mytests, scenario, name, numevents, command, dataset, inputFilesConfig):
+    # activate these when the requisite MC exists
```
Comment (kpedro88) [inline Production/test/unitTest.py]: this comment can probably be deleted, it's very outdated

File: setup.sh [hunk 3]
```diff
@@ -15,6 +16,7 @@ usage(){
 	echo ""
 	echo "-f [fork]           clone from specified fork (default = ${FORK})"
 	echo "-b [branch]         clone specified branch (default = ${BRANCH})"
+    echo "-B [batch]          configure some settings for checkout within batch setups (default = ${BATCH})"
```
Comment (kpedro88) [inline setup.sh]: use tabs not spaces

File: setup.sh [hunk 4]
```diff
@@ -26,12 +28,14 @@ usage(){
...
 
 # process options
-while getopts "f:b:a:j:n:d:c:h" opt; do
+while getopts "f:b:Ba:j:n:d:c:h" opt; do
 	case "$opt" in
 	f) FORK=$OPTARG
 	;;
 	b) BRANCH=$OPTARG
 	;;
+    B) BATCH=--upstream-only
```
Comment (kpedro88) [inline setup.sh]: use tabs not spaces

File: setup.sh [hunk 5]
```diff
@@ -60,6 +64,22 @@ else
 	usage 1
 fi
 
+# make sure that scram is available
+# examples:
+#   in a non-interactive Docker container this will evaluate to "hB"
+#   in an login session this will evaluate to "himBCHP"
```
Comment (kpedro88) [inline setup.sh]: typo "an login" -> "a login"

File: setup.sh [hunk 6]
```diff
@@ -60,6 +64,22 @@ else
...
+if [[ $- != *i* ]]; then
+	echo "WARNING: Non-interactive shell found!"
+    if [[ -f "${HOME}/.bashrc" ]]; then
```
Comment (kpedro88) [inline setup.sh]: use tabs not spaces (everywhere in this file)

File: setup.sh (hunk 3 above)
Comment (kpedro88) [reply setup.sh]: also add this option to the readme

File: setup.sh [hunk 7]
```diff
@@ -94,9 +123,16 @@ cd ${DIR}/${NAME}/src
 
 # cmsenv
 eval `scramv1 runtime -sh`
-git cms-init $ACCESS_CMSSW
+git cms-init $ACCESS_CMSSW $BATCH
 git config gc.auto 0
 
+# batch git config
+if [[ ! -z "$BATCH" ]]; then
```
Comment (kpedro88) [inline setup.sh]: `-n` is a little clearer than `! -z` (to me, anyway)

File: Production/test/unitTest.py [hunk 8]
```diff
@@ -1,92 +1,157 @@
...
+    if test==-1: name = ""
+
+    # list of tests
+    mytests = []
+    defineTests(mytests,scenario,name,numevents,command,dataset,inputFilesConfig)
+
+    # sanity check for defining an on-the-fly test
+    if test==len(mytests) and dataset=="" and inputFilesConfig=="":
+        print "If defining a unitTest on-the-fly, you must either specify a \'dataset\' or an \'inputFilesConfig\'."
+        return -1
```
Comment (kpedro88) [inline Production/test/unitTest.py]: this return seems to be discarded, should it be emitted as an exit code?

File: .github/workflows/dockerimage.yml (hunk 1 above)
Comment (aperloff) [reply .github/workflows/dockerimage.yml:32]: Done

File: Production/test/unitTest.py (hunk 2 above)
Comment (aperloff) [reply Production/test/unitTest.py]: Done

File: setup.sh (hunk 3 above)
Comment (aperloff) [reply setup.sh]: Done

File: setup.sh (hunk 5 above)
Comment (aperloff) [reply setup.sh]: Fixed

File: setup.sh (hunk 4 above)
Comment (aperloff) [reply setup.sh]: Fixed

File: setup.sh (hunk 6 above)
Comment (aperloff) [reply setup.sh]: Fixed

File: setup.sh (hunk 7 above)
Comment (aperloff) [reply setup.sh]: Switched

File: Production/test/unitTest.py (hunk 8 above)
Comment (aperloff) [reply Production/test/unitTest.py]: Fixed

File: Production/test/unitTest.py [hunk 9]
```diff
@@ -1,92 +1,156 @@
...
+    if test==-1: name = ""
+
+    # list of tests
+    mytests = []
+    defineTests(mytests,scenario,name,numevents,command,dataset,inputFilesConfig)
+
+    # sanity check for defining an on-the-fly test
+    if test==len(mytests) and dataset=="" and inputFilesConfig=="":
+        print "If defining a unitTest on-the-fly, you must either specify a \'dataset\' or an \'inputFilesConfig\'."
+        exit(-1)
```
Comment (kpedro88) [inline Production/test/unitTest.py]: best practice is `sys.exit`

Comment (aperloff) [reply Production/test/unitTest.py]: Changed

**Pull request merged**